daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -44,13 +45,81 @@ ipa_kdcpolicy_check_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
                        const char **status, krb5_deltat *lifetime_out,
                        krb5_deltat *renew_lifetime_out)
 {
+    krb5_error_code kerr;
+    enum ipadb_user_auth ua;
+    struct ipadb_e_data *ied;
+    int valid_auth_indicators = 0;
+
     *status = NULL;
-    *lifetime_out = 0;
-    *renew_lifetime_out = 0;
 
     krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
 
-    return 0;
+    ied = (struct ipadb_e_data *)client->e_data;
+    if (ied->magic != IPA_E_DATA_MAGIC) {
+        /* e-data is not availble, getting user auth from LDAP */
+        krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: client e_data not availble. Try fetching...");
+        kerr = ipadb_get_principal(context, request->client, KRB5_KDB_FLAG_ALIAS_OK, &client);
+        if (kerr != 0) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_find_principal failed.");
+            goto done;
+        }
+
+        ied = (struct ipadb_e_data *)client->e_data;
+        if (ied->magic != IPA_E_DATA_MAGIC) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: client e_data fetching failed.");
+            goto done;
+        }
+    }
+
+    ua = ied->user_auth;
+    
+    /* For each auth indicator, see if it is allowed for that user */
+    for (int i = 0; auth_indicators[i] != NULL; i++) {
+        const char *auth_indicator = auth_indicators[i];
+
+        if (strcmp(*auth_indicators, "otp") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_OTP)) {
+                kerr = KRB5KDC_ERR_POLICY;
+                *status = "OTP pre-authentication not allowed for this user.";
+                goto done;
+            }
+
+        } else if (strcmp(*auth_indicators, "radius") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_RADIUS)) {
+                kerr = KRB5KDC_ERR_POLICY;
+                *status = "RADIUS pre-authentication not allowed for this user.";
```

Comment:
Radius is an internal detail of the way OTP can be configured, I think we should still return "OTP pre-authentication ..." as the status error that is sent to the client ? Or is this error only sent to logs ?

Reply:
I didn't see the message in krb5 packet returned to client, so I guess this is only showing up in the logs.